### PR TITLE
>3-5x faster torch.compile forward compilation for autoregressive decoder models

### LINKED ISSUE #29165: [`generate` + static cache + `torch.compile`] ability to pass statically shaped 4D `attention_mask` to the model forward
### Feature request

Currently, the `attention_mask` passed is 2D and of dynamic shapes.

This causes issues when using a compiled model with `model.forward = torch.compile(model.forward, mode="reduce-overhead")`, see https://github.com/pytorch/pytorch/issues/120309 & https://github.com/huggingface/transformers/pull/29114.

```
----- in forward 0
name=input_ids, shape=torch.Size([2, 7]), stride=(7, 1), dtype=torch.int64, device=cuda:0
name=position_ids, shape=torch.Size([2, 7]), stride=(7, 1), dtype=torch.int64, device=cuda:0
name=cache_position, shape=torch.Size([7]), stride=(1,), dtype=torch.int64, device=cuda:0
name=past_key_values, value=None
name=use_cache, value=True
name=attention_mask, shape=torch.Size([2, 7]), stride=(7, 1), dtype=torch.int64, device=cuda:0
forward call latency: 1784.737 ms      <---------------------------- EXTREMELY SLOW.
----- in forward 1
name=input_ids, shape=torch.Size([2, 1]), stride=(1, 1), dtype=torch.int64, device=cuda:0
name=position_ids, shape=torch.Size([2, 1]), stride=(1, 1), dtype=torch.int64, device=cuda:0
name=cache_position, shape=torch.Size([1]), stride=(1,), dtype=torch.int64, device=cuda:0
name=past_key_values, value=None
name=use_cache, value=True
name=attention_mask, shape=torch.Size([2, 8]), stride=(8, 1), dtype=torch.int64, device=cuda:0
forward call latency: 1851.579 ms      <---------------------------- EXTREMELY SLOW.
----- in forward 2
name=input_ids, shape=torch.Size([2, 1]), stride=(1, 1), dtype=torch.int64, device=cuda:0
name=position_ids, shape=torch.Size([2, 1]), stride=(1, 1), dtype=torch.int64, device=cuda:0
name=cache_position, shape=torch.Size([1]), stride=(1,), dtype=torch.int64, device=cuda:0
name=past_key_values, value=None
name=use_cache, value=True
name=attention_mask, shape=torch.Size([2, 9]), stride=(9, 1), dtype=torch.int64, device=cuda:0
forward call latency: 1421.504 ms      <---------------------------- EXTREMELY SLOW.
----- in forward 3
name=input_ids, shape=torch.Size([2, 1]), stride=(1, 1), dtype=torch.int64, device=cuda:0
name=position_ids, shape=torch.Size([2, 1]), stride=(1, 1), dtype=torch.int64, device=cuda:0
name=cache_position, shape=torch.Size([1]), stride=(1,), dtype=torch.int64, device=cuda:0
name=past_key_values, value=None
name=use_cache, value=True
name=attention_mask, shape=torch.Size([2, 10]), stride=(10, 1), dtype=torch.int64, device=cuda:0
forward call latency: 1740.283 ms      <---------------------------- EXTREMELY SLOW.
```

Instead, we may want to pass directly 4D masks of static shape to the model, avoiding cuda graphs recomputation. This may allow the compile time to go down from >3 min to a mere 50s.

### Motivation

/

### Your contribution

/

### LINKED ISSUE #32313: HybridCache slow after reset
### System Info

- `transformers` version: 4.44.0.dev0                                                                                   
- Platform: Linux-5.4.0-166-generic-x86_64-with-glibc2.31                                                               
- Python version: 3.11.9                                                                                                
- Huggingface_hub version: 0.23.4  
- Safetensors version: 0.4.3
- Accelerate version: 0.30.0
- Accelerate config:    - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: bf16
        - use_cpu: False
        - debug: False
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: 0,1
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - enable_cpu_affinity: False
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
- PyTorch version (GPU?): 2.3.1+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA A100-SXM4-80GB

### Who can help?

@sanchit-gandhi @gante @ArthurZucker 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import os
os.environ["TOKENIZERS_PARALLELISM"] = "false"

import time
from transformers import AutoTokenizer, Gemma2ForCausalLM
from transformers.cache_utils import HybridCache
import torch

torch.set_float32_matmul_precision("high")
# catch re-compilations
torch._logging.set_logs(graph_breaks=True, recompiles=True)

tokenizer = AutoTokenizer.from_pretrained("google/gemma-2-9b")
model = Gemma2ForCausalLM.from_pretrained("google/gemma-2-9b", attn_implementation="eager")
model.to("cuda")

model.forward = torch.compile(model.forward, mode="reduce-overhead", fullgraph=True)

input_text = "The theory of special relativity states "
input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
prompt_length = input_ids.input_ids.shape[1]

model.generation_config.min_new_tokens = model.generation_config.max_new_tokens = 32

past_key_values = HybridCache(
    config=model.config,
    max_batch_size=1,
    max_cache_len=prompt_length + 4 * model.generation_config.max_new_tokens,
    device=model.device,
    dtype=model.dtype
)

# enable passing kv cache
model._supports_cache_class = True
model.generation_config.cache_implementation = None

for i in range(3):
    # two warm-ups
    outputs_1 = model.generate(**input_ids, past_key_values=past_key_values, do_sample=True, temperature=1)
    outputs_2 = model.generate(outputs_1, past_key_values=past_key_values, do_sample=True, temperature=1)

    # one timed run
    torch.cuda.synchronize("cuda")
    start = time.time()
    outputs_3 = model.generate(outputs_2, past_key_values=past_key_values, do_sample=True, temperature=1)
    torch.cuda.synchronize("cuda")
    runtime = time.time() - start
    print(f"Run {i}: {model.generation_config.max_new_tokens / runtime} tok/s")

    past_key_values.reset()
```

**Print Output:**
```
V0730 05:56:47.345000 140167293411712 torch/_dynamo/guards.py:1425] [__recompiles] Recompiling function forward in /home/sanchit/transformers/src/transformers/models/gemma2/modeling_gemma2.py:891
V0730 05:56:47.345000 140167293411712 torch/_dynamo/guards.py:1425] [__recompiles]     triggered by the following guard failure(s):
V0730 05:56:47.345000 140167293411712 torch/_dynamo/guards.py:1425] [__recompiles]     - tensor 'L['input_ids']' stride mismatch at index 0. expected 8, actual 1

bdV0730 05:57:41.549000 140167293411712 torch/_dynamo/guards.py:1425] [__recompiles] Recompiling function forward in /home/sanchit/transformers/src/transformers/models/gemma2/modeling_gemma2.py:891
V0730 05:57:41.549000 140167293411712 torch/_dynamo/guards.py:1425] [__recompiles]     triggered by the following guard failure(s):
V0730 05:57:41.549000 140167293411712 torch/_dynamo/guards.py:1425] [__recompiles]     - tensor 'L['input_ids']' stride mismatch at index 0. expected 1, actual 40
V0730 05:57:41.549000 140167293411712 torch/_dynamo/guards.py:1425] [__recompiles]     - tensor 'L['input_ids']' stride mismatch at index 0. expected 8, actual 40

Run 0: 28.8159080589 tok/s
Run 1: 0.878302057247666 tok/s
Run 2: 19.946942197324718 tok/s
```

=> we get only two recompilations (expected), but the inference speed of the second and third runs are significantly lower than the first. This pattern happens only after calling `past_key_values.reset()`, which suggests a bug in how we're resetting the `HybridCache`.

### Expected behavior

```
Run 0: 28.8159080589 tok/s
Run 1: 28.8159080589 tok/s
Run 2: 28.8159080589 tok/s
```

### PR DESCRIPTION
**Pull request by fxmarty**
As detailed in https://github.com/huggingface/transformers/pull/29114 and https://github.com/huggingface/transformers/issues/29165, the current implementation of `GenerationMixin.generate` passes 2D integer tensors of variable length as `attention_mask` to the models forward, that are transformed into 4D tensors of static shape in the models forward for decoder models.

This is an issue when using torch.compile with static cache, as then the model forward has inputs of various shape at each decoding step, which retriggers a cuda graph capture. This is especially bad when setting large enough `max_new_tokens` (>64) as cuda graphs are recaptured many many times.

Effectively as per https://pytorch.org/docs/stable/torch.compiler_cudagraph_trees.html#cudagraph-trees-integration, what is to be considered the compilation time is the `torch.compile` call, and the subsequent two forward calls (without guard invalidation, cudagraph invalidation).

For example, below, we set `max_new_tokens = 256`, with an 8B model on an A100, and on main the compilation using `model.forward = torch.compile(model.forward, mode="reduce-overhead", fullgraph=True)` takes:
```
----- GENERATE WITHOUT COMPILE
Non-compiled generate call took (no warmup): 12.546 s
decoded ['I would like (redacted)"]
compiling...
Finished compile call: 0.926 s
----- GENERATE WITH COMPILE
/home/felix/miniconda3/envs/fx/lib/python3.9/site-packages/torch/_inductor/compile_fx.py:150: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
First generate call took: 89.536 s (the first run warms up each graph, which does things like CuBlas or Triton benchmarking)
-------- decoded 1 ['I would like (redacted)"]
Second generate call took: 267.591 s (The second run does a CUDA Graph recording, and replays it)
-------- decoded 2 ['I would like (redacted)"]
Third generate call took: 3.421 s (Finally we hit the optimized, CUDA Graph replay path)
-------- decoded 3 ['I would like (redacted)"]

```

while with this PR the compilation takes
```
----- GENERATE WITHOUT COMPILE
Non-compiled generate call took (no warmup): 11.672 s
decoded ['I would like (redacted)"]
compiling...
Finished compile call: 0.999 s
----- GENERATE WITH COMPILE
First generate call took: 85.505 s (the first run warms up each graph, which does things like CuBlas or Triton benchmarking)
-------- decoded 1 ['I would like (redacted)"]
Second generate call took: 4.845 s (The second run does a CUDA Graph recording, and replays it)
-------- decoded 2 ['I would like (redacted)"]
Third generate call took: 3.495 s (Finally we hit the optimized, CUDA Graph replay path)
-------- decoded 3 ['I would like (redacted)"]
```

repro here https://gist.github.com/fxmarty/74b4463935ada44ef7755a544cae8773, using `dynamic=True` is even worse

cc @LysandreJik @ArthurZucker @amyeroberts @gante I think we'll need a lot of cleanup in `modeling_attn_mask_utils`  at some point - we have many `_prepare_attention_mask_xxx` functions (even some in `PreTrainedModel`)

Closes https://github.com/huggingface/transformers/issues/29165

### CODE REVIEW

File: src/transformers/models/mistral/modeling_mistral.py [hunk 1]
```diff
@@ -1065,7 +1065,6 @@ def forward(
             attentions=outputs.attentions,
         )
 
-    # Copied from transformers.models.llama.modeling_llama.LlamaForCausalLM.prepare_inputs_for_generation
```
Comment (fxmarty) [inline src/transformers/models/mistral/modeling_mistral.py:1068]: might check it with sliding window

**Review by ArthurZucker — COMMENTED** [review]
Very very nice work! The fix ends up being quite "simple". 
2 things:
1. In accordance to `transformers`-philosophy, the causal mask creation should stay in the modeling file as it is CORE. Wrapping it in a function is not a problem !
2. This attention mask 4D created in `prepare_input_for_generation` should be done in `generate` automatically, but that is a refactor for another PR cc @gante and @zucchini-nlp as this is in your roadmap. 

Let's just copy past the function in each modeling code, maybe a comment saying why we copy past it (it's core and does not need to be hidden away, unlike that call to fa2 for example)

**Comment by fxmarty** [comment]
> In accordance to transformers-philosophy, the causal mask creation should stay in the modeling file as it is CORE. Wrapping it in a function is not a problem !

Understood.

> This attention mask 4D created in prepare_input_for_generation should be done in generate automatically

I am not sure about this, as it is model specific (some models support `cache_position` & static cache, others don't). Do you mean that the 2D -> 4D update should preferably be handled in `_update_model_kwargs_for_generation` instead of `prepare_inputs_for_generation`?

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -1216,11 +1260,36 @@ def prepare_inputs_for_generation(
...
+                position_ids = position_ids.clone(memory_format=torch.contiguous_format)
+
         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
         if inputs_embeds is not None and cache_position[0] == 0:
             model_inputs = {"inputs_embeds": inputs_embeds}
         else:
-            model_inputs = {"input_ids": input_ids.contiguous()}  # `contiguous()` needed for compilation use cases
+            model_inputs = {"input_ids": input_ids}
+
+        if isinstance(past_key_values, StaticCache) and attention_mask.ndim == 2:
```
Comment (gante) [inline src/transformers/models/llama/modeling_llama.py:1272]: We need a TODO for me/Arthur here :D 

TL;DR `StaticCache` is not the only compileable cache. At the moment we don't have an attribute that says "this cache supports compilation of the fwd pass". We should create that attribute and use it here

**Review by gante — APPROVED** [review]
🔥 🔥 🔥 This is great

**Comment by fxmarty** [comment]
I'll add some tests here.

File: src/transformers/models/llama/modeling_llama.py (hunk 2 above)
Comment (sanchit-gandhi) [reply src/transformers/models/llama/modeling_llama.py:1272]: +1 on this! Currently, our check for Gemma-2 is looking at:
```python
       if isinstance(past_key_values, HybridCache) and attention_mask.ndim == 2:
```

=> it would be great to remove the ambiguity and truly make this a cache agnostic function

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -1216,11 +1260,36 @@ def prepare_inputs_for_generation(
             if past_key_values:
                 position_ids = position_ids[:, -input_ids.shape[1] :]
 
+                # This `clone` call is needed to avoid recapturing cuda graphs with `torch.compile`'s  `mode="reduce-overhead`, as otherwise the input `position_ids` would have various stride during the decoding. Here, simply using `.contiguous()` is not sufficient as in the batch size = 1 case, `position_ids` is already contiguous but with varying stride which retriggers a capture.
+                position_ids = position_ids.clone(memory_format=torch.contiguous_format)
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:1264]: nice catch

File: src/transformers/models/llama/modeling_llama.py (hunk 2 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:1272]: I think we should create the attention mask once indeed, and just "update" it to unpad after each decoding step (pay attention) , if possible in generate, this way we don't change prepare_inputs_for_generation. We can do better checks of the Cache class in generate as well

**Review by ArthurZucker — COMMENTED** [review]
LGTM, IMO we should create the mask in generate, but otherwise mergeable

**Comment by fxmarty** [comment]
I checked the slow tests. There are no other tests failing due to this PR. Although on `main` there are a few tests failing unrelated to this PR which may be concerning (on A100, maybe some reference are for an other GPU, especially gemma):

```
tests/models/cohere/test_modeling_cohere.py::CohereModelTest::test_generate_compile_fullgraph
tests/models/gemma/test_modeling_gemma.py::GemmaModelTest::test_generate_compile_fullgraph
tests/models/gemma/test_modeling_gemma.py::GemmaIntegrationTest::test_model_2b_bf16
tests/models/gemma/test_modeling_gemma.py::GemmaIntegrationTest::test_model_2b_bf16_dola
tests/models/gemma/test_modeling_gemma.py::GemmaIntegrationTest::test_model_2b_eager
tests/models/gemma/test_modeling_gemma.py::GemmaIntegrationTest::test_model_2b_sdpa
tests/models/gemma/test_modeling_gemma.py::GemmaIntegrationTest::test_model_7b_4bit
tests/models/gemma/test_modeling_gemma.py::GemmaIntegrationTest::test_model_7b_bf16
tests/models/gemma2/test_modeling_gemma2.py::GemmaModelTest::test_generate_compile_fullgraph
tests/models/gemma2/test_modeling_gemma2.py::Gemma2ModelTest::test_sdpa_can_dispatch_on_flash
tests/models/gemma2/test_modeling_gemma2.py::Gemma2ModelTest::test_sdpa_equivalence
tests/models/llama/test_modeling_llama.py::LlamaModelTest::test_generate_compile_fullgraph
tests/models/llama/test_modeling_llama.py::LlamaIntegrationTest::test_model_7b_logits
tests/models/llama/test_modeling_llama.py::LlamaIntegrationTest::test_model_7b_logits_bf16
``` 

Testing for `mistral or mixtral or olmo or persimmon or phi or phi3 or qwen2 or qwen2_moe or stablelm or starcoder2 or whisper` now.
